File: lang/kotlin/src/main/java/com/abi/Mobile.kt
package com.abi

/**
 * Sensor data returned by [MobileContext.readSensor].
 */
data class SensorData(
    val timestampMs: Long,
    val values: FloatArray,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SensorData) return false
        return timestampMs == other.timestampMs && values.contentEquals(other.values)
    }

    override fun hashCode(): Int {
        var result = timestampMs.hashCode()
        result = 31 * result + values.contentHashCode()
        return result
    }
}

/**
 * Device information returned by [MobileContext.getDeviceInfo].
 */
data class DeviceInfo(
    val screenWidth: Int,
    val screenHeight: Int,
    val batteryLevel: Float,
    val isCharging: Boolean,
    val platform: String,
    val osVersion: String,
    val deviceModel: String,
)

/** Sensor type constants. */
object SensorType {
    const val ACCELEROMETER = 0
    const val GYROSCOPE     = 1
    const val MAGNETOMETER  = 2
    const val GPS           = 3
    const val BAROMETER     = 4
    const val PROXIMITY     = 5
    const val LIGHT         = 6
}

/** Permission type constants. */
object PermissionType {
    const val CAMERA        = 0
    const val MICROPHONE    = 1
    const val LOCATION      = 2
    const val NOTIFICATIONS = 3
    const val STORAGE       = 4
    const val CONTACTS      = 5
    const val BLUETOOTH     = 6
}

/** Permission status constants. */
object PermissionStatus {
    const val GRANTED       = 0
    const val DENIED        = 1
    const val NOT_REQUESTED = 2
}

/** Notification priority constants. */
object NotificationPriority {
    const val LOW      = 0
    const val NORMAL   = 1
    const val HIGH     = 2
    const val CRITICAL = 3
}

/**
 * Kotlin wrapper around the ABI mobile C API via JNI.
 *
 * Create an instance with [MobileContext] and call [close] when finished.
 * Implements [AutoCloseable] so it can be used with `use {}`.
 */
class MobileContext : AutoCloseable {
    /** Native pointer held by JNI. */
    private var nativePtr: Long = 0L

    init {
        nativePtr = nativeInit()
        if (nativePtr == 0L) {
            throw RuntimeException("Failed to initialize MobileContext")
        }
    }

    /** Read a sensor value. */
    fun readSensor(sensorType: Int): SensorData = nativeReadSensor(nativePtr, sensorType)

    /** Send a notification. */
    fun sendNotification(
        title: String,
        body: String,
        priority: Int = NotificationPriority.NORMAL,
    ): Int = nativeSendNotification(nativePtr, title, body, priority)

    /** Get device information. */
    fun getDeviceInfo(): DeviceInfo = nativeGetDeviceInfo(nativePtr)

    /** Check the status of a permission. */
    fun checkPermission(permission: Int): Int = nativeCheckPermission(nativePtr, permission)

    /** Request a permission. */
    fun requestPermission(permission: Int): Int = nativeRequestPermission(nativePtr, permission)

    /** Number of tracked notifications. */
    val notificationCount: Int get() = nativeGetNotificationCount(nativePtr)

    /** Clear all tracked notifications. */
    fun clearNotifications() = nativeClearNotifications(nativePtr)

    override fun close() {
        if (nativePtr != 0L) {
            nativeDestroy(nativePtr)
            nativePtr = 0L
        }
    }

    // JNI native methods
    private external fun nativeInit(): Long
    private external fun nativeDestroy(ptr: Long)
    private external fun nativeReadSensor(ptr: Long, sensorType: Int): SensorData
    private external fun nativeSendNotification(ptr: Long, title: String, body: String, priority: Int): Int
    private external fun nativeGetDeviceInfo(ptr: Long): DeviceInfo
    private external fun nativeCheckPermission(ptr: Long, permission: Int): Int
    private external fun nativeRequestPermission(ptr: Long, permission: Int): Int
    private external fun nativeGetNotificationCount(ptr: Long): Int
    private external fun nativeClearNotifications(ptr: Long)

    companion object {
        init {
            System.loadLibrary("abi_jni")
        }
    }
}

File: lang/swift/Sources/ABI/Mobile.swift
import CABI

/// Idiomatic Swift wrapper around the ABI mobile C API.
public final class MobileContext {
    private var handle: OpaquePointer?

    /// Initialize a mobile context.
    public init() throws {
        var ptr: OpaquePointer?
        let result = abi_mobile_init(&ptr)
        guard result == ABI_OK else {
            throw ABIError(code: result)
        }
        self.handle = ptr
    }

    deinit {
        if let h = handle {
            abi_mobile_destroy(h)
        }
    }

    // MARK: - Sensors

    /// Sensor type constants matching the C API.
    public enum SensorType: Int32 {
        case accelerometer  = 0
        case gyroscope      = 1
        case magnetometer   = 2
        case gps            = 3
        case barometer      = 4
        case proximity      = 5
        case light          = 6
    }

    /// Data returned from a sensor reading.
    public struct SensorData {
        public let timestampMs: UInt64
        public let values: (Float, Float, Float)
    }

    /// Read a sensor value.
    public func readSensor(_ type: SensorType) throws -> SensorData {
        var raw = abi_sensor_data_t()
        let result = abi_mobile_read_sensor(handle, type.rawValue, &raw)
        guard result == ABI_OK else {
            throw ABIError(code: result)
        }
        return SensorData(
            timestampMs: raw.timestamp_ms,
            values: (raw.values.0, raw.values.1, raw.values.2)
        )
    }

    // MARK: - Notifications

    /// Notification priority levels.
    public enum NotificationPriority: Int32 {
        case low      = 0
        case normal   = 1
        case high     = 2
        case critical = 3
    }

    /// Send a notification.
    public func sendNotification(title: String, body: String,
                                 priority: NotificationPriority = .normal) throws {
        let result = abi_mobile_send_notification(handle, title, body, priority.rawValue)
        guard result == ABI_OK else {
            throw ABIError(code: result)
        }
    }

    /// The number of tracked notifications.
    public var notificationCount: Int {
        Int(abi_mobile_get_notification_count(handle))
    }

    /// Clear all tracked notifications.
    public func clearNotifications() {
        abi_mobile_clear_notifications(handle)
    }

    // MARK: - Device Info

    /// Device information snapshot.
    public struct DeviceInfo {
        public let screenWidth: UInt32
        public let screenHeight: UInt32
        public let batteryLevel: Float
        public let isCharging: Bool
        public let platform: String
        public let osVersion: String
        public let deviceModel: String
    }

    /// Get current device information.
    public func getDeviceInfo() throws -> DeviceInfo {
        var raw = abi_device_info_t()
        let result = abi_mobile_get_device_info(handle, &raw)
        guard result == ABI_OK else {
            throw ABIError(code: result)
        }
        return DeviceInfo(
            screenWidth: raw.screen_width,
            screenHeight: raw.screen_height,
            batteryLevel: raw.battery_level,
            isCharging: raw.is_charging,
            platform: String(cString: raw.platform),
            osVersion: String(cString: raw.os_version),
            deviceModel: String(cString: raw.device_model)
        )
    }

    // MARK: - Permissions

    /// Permission types.
    public enum Permission: Int32 {
        case camera        = 0
        case microphone    = 1
        case location      = 2
        case notifications = 3
        case storage       = 4
        case contacts      = 5
        case bluetooth     = 6
    }

    /// Permission status values.
    public enum PermissionStatus: Int32 {
        case granted      = 0
        case denied       = 1
        case notRequested = 2
    }

    /// Check the current status of a permission.
    public func checkPermission(_ permission: Permission) -> PermissionStatus {
        let raw = abi_mobile_check_permission(handle, permission.rawValue)
        return PermissionStatus(rawValue: raw) ?? .notRequested
    }

    /// Request a permission.
    @discardableResult
    public func requestPermission(_ permission: Permission) -> PermissionStatus {
        let raw = abi_mobile_request_permission(handle, permission.rawValue)
        return PermissionStatus(rawValue: raw) ?? .notRequested
    }
}
